server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1634,116 +1660,40 @@ public Entitlement adjustEntitlementQuantity(Consumer consumer, Entitlement enti
         Map<String, Integer> poolQuantityMap, CallerType caller)
         throws EntitlementRefusedException {
 
-        // Because there are several paths to this one place where entitlements
-        // are granted, we cannot be positive the caller obtained a lock on the
-        // pool when it was read. As such we're going to reload it with a lock
-        // before starting this process.
-        log.debug("Locking pools: {}", poolQuantityMap.keySet());
-        Collection<Pool> pools = poolCurator.lockAndLoadByIds(poolQuantityMap.keySet());
-
-        if (log.isDebugEnabled()) {
-            for (Pool pool : pools) {
-                log.debug("Locked pool: {} consumed: {}", pool, pool.getConsumed());
-            }
-        }
-
-        log.debug("Done locking pools");
+        BindChain chain = new BindChain();
+        BindContext context = bindContextFactory.create(consumer, poolQuantityMap);
+        chain.addOperation(rulesCheckOpFactory.create(caller));
+        chain.addOperation(handleEntitlementsOp);
+        chain.addOperation(postBindBonusPoolsOp);
+        chain.addOperation(handleCertificatesOp);
+        chain.addOperation(complianceOp);
 
-        Map<String, PoolQuantity> poolQuantities = new HashMap<String, PoolQuantity>();
-        boolean quantityFound = false;
-        for (Pool pool : pools) {
-            Integer quantity = poolQuantityMap.get(pool.getId());
-            if (quantity > 0) {
-                quantityFound = true;
-            }
-
-            poolQuantities.put(pool.getId(), new PoolQuantity(pool, quantity));
-            poolQuantityMap.remove(pool.getId());
+        // fetch and pre-process entities before locks are placed.
+        chain.preProcess(context);
+        if (context.getException() == null) {
+            context.lockPools();
+            // post lock operations
+            chain.execute(context);
```

Comment:
This looks like it'll only run whatever operation happens to be in slot 1. since each call to .preprocess and .execute only grabs the current index and executes it iff the execute operation count hasn't exceeded the operation count.

At the time of writing, this seems counterintuitive. I'll update this comment if I see something else that clears up why this currently works the way it does.